### PR TITLE
sdk-base: use updated account data for processing direct rooms

### PR DESCRIPTION
**Pull request by jmartinesp**
When processing a sliding sync response, if no account_data for direct rooms is found in the current response, use the cached value in the store.

This fixes an issue in the EX clients where only very few DMs were visible under the `People` filter since they weren't considered DMs, until a full restart of the app.

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-base/src/client.rs
Comment (poljar) [inline crates/matrix-sdk-base/src/client.rs:1172]: I don't think that this is the correct place to do this. Do note that the currently received account data will only be persisted after the following code snippet further bellow:

https://github.com/matrix-org/matrix-rust-sdk/blob/5827bb7ab33b275835f16eacb4c0fecfa79f4a2b/crates/matrix-sdk-base/src/client.rs#L1187-L1188

If the current sync has some new account data, this `handle_account_data()` call will overwrite it with the old one found in the store.

I think that the most correct way would be to apply the direct targets when we initially create the `Room` object, though that might mean that we load the account data from the store multiple times. I can't come up with a more performant way for this right now.

**Review by poljar — CHANGES_REQUESTED** [review]
I think that this introduces a subtle bug where we might go back in time with the account data.

File: crates/matrix-sdk-base/src/client.rs
Comment (poljar) [inline crates/matrix-sdk-base/src/client.rs:1172]: Still not correct, consider the following scenario:

We get some updated account data events, but the `m.direct` event was not updated. So `response.account_data` will not be empty, but it will also not contain the `m.direct` event. Thus rooms will remain without their direct target.

**Review by poljar — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-base/src/client.rs
Comment (jmartinesp) [reply crates/matrix-sdk-base/src/client.rs]: I hope https://github.com/matrix-org/matrix-rust-sdk/pull/3990/commits/b53fa163922365d0ec021c4b2a767b7ff3d6e02b handles that in a better way 🤞 .

**Review by poljar — APPROVED** [review]
Ok, this can go in as a fix. If you have the time and motivation it would be good to get rid of this TODO item above as well.

For this we would split the `handle_account_data()` method into two methods, one that handles all account data except for `m.direct` and one that handles just the `m.direct` events.

I think then this PR also becomes less of a hack.

**Comment by jmartinesp** [comment]
Since we want to merge this ASAP, do you mind if we merge this as is, then I immediately start working in a PR with the changes you propose? That way we have the fix in now since we need, and we can have the right structure soon-ish (today, tomorrow?).

**Comment by poljar** [comment]
> Since we want to merge this ASAP, do you mind if we merge this as is, then I immediately start working in a PR with the changes you propose? That way we have the fix in now since we need it, and we can have the right structure soon-ish (today, tomorrow?).

I don't mind this going in, no. After all I approved the PR and said as much in my review:

> Ok, this can go in as a fix.

**Pull request merged**
